File: PriceAndDistanceLogic.java
// ==========================================
// BUSINESS LOGIC REQUESTED BY USER
// ==========================================

import java.util.*;

public class PriceAndDistanceLogic {

    // 1. Weather-Impact Algorithm for Smart Market
    public static class CropPricing {
        String cropName;
        double currentPrice;
        
        public CropPricing(String name, double price) {
            this.cropName = name;
            this.currentPrice = price;
        }

        /**
         * Calculates predicted price based on weather parameters.
         * @param rainfallMm Expected rainfall in mm (e.g. 50.0)
         * @param tempCelsius Expected temperature in Celsius
         * @param isDroughtWarning boolean
         * @return predicted price per Quintal
         */
        public double calculatePredictedPrice(double rainfallMm, double tempCelsius, boolean isDroughtWarning) {
            double impactFactor = 1.0;

            // Simple Logic for demonstration
            if (isDroughtWarning) {
                // Drought increases prices significantly due to scarcity
                impactFactor += 0.15; 
            } else if (rainfallMm > 150) {
                // Heavy rains might damage crops, increasing prices
                impactFactor += 0.05;
            } else if (rainfallMm > 40 && rainfallMm <= 150) {
                // Optimal rain provides good harvest, slight decrease in price
                impactFactor -= 0.05;
            }

            // Extreme temperatures affect yield
            if (tempCelsius > 40.0 || tempCelsius < 10.0) {
                impactFactor += 0.08;
            }

            return this.currentPrice * impactFactor;
        }
    }

    // 2. Distance Sorting Logic for Storage Locator
    public static class StorageUnit {
        String name;
        double latitude;
        double longitude;

        public StorageUnit(String name, double lat, double lon) {
            this.name = name;
            this.latitude = lat;
            this.longitude = lon;
        }
    }

    /**
     * Haversine formula to calculate the distance between two points on the earth
     */
    public static double haversineCalculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final int R = 6371; // Radious of the earth in km
        double latDistance = Math.toRadians(lat2 - lat1);
        double lonDistance = Math.toRadians(lon2 - lon1);
        
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2) + 
                   Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) * 
                   Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
                   
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }

    /**
     * Sorts list of Cold Storage units based on distance from the Farmer
     */
    public static List<StorageUnit> getSortedStorageLocators(double farmerLat, double farmerLon, List<StorageUnit> storageUnits) {
        storageUnits.sort(Comparator.comparingDouble(
            unit -> haversineCalculateDistance(farmerLat, farmerLon, unit.latitude, unit.longitude)
        ));
        return storageUnits;
    }

    public static void main(String[] args) {
        // Example Usage
        CropPricing wheat = new CropPricing("Wheat", 2200.0);
        double predicted = wheat.calculatePredictedPrice(160, 35.0, false);
        System.out.println("Predicted Wheat Price: " + predicted);

        List<StorageUnit> storages = new ArrayList<>();
        storages.add(new StorageUnit("Kisan Storage", 28.7041, 77.1025)); 
        storages.add(new StorageUnit("AgriCold", 28.5355, 77.3910));
        
        // Output nearest storage locations
        List<StorageUnit> sorted = getSortedStorageLocators(28.6139, 77.2090, storages);
        for(StorageUnit s : sorted) {
            System.out.println("Storage: " + s.name);
        }
    }
}

File: PriceAndDistanceLogic.cpp
// ==========================================
// BUSINESS LOGIC REQUESTED BY USER
// ==========================================

#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

using namespace std;

// 1. Weather-Impact Algorithm for Aaj Ka Bhav
class CropPricing {
public:
    string cropName;
    double currentPrice;

    CropPricing(string name, double price) : cropName(name), currentPrice(price) {}

    // Calculates predicted price based on weather parameters
    double calculatePredictedPrice(double rainfallMm, double tempCelsius, bool isDroughtWarning) {
        double impactFactor = 1.0;

        if (isDroughtWarning) {
            impactFactor += 0.15; 
        } else if (rainfallMm > 150) {
            impactFactor += 0.05;
        } else if (rainfallMm > 40 && rainfallMm <= 150) {
            impactFactor -= 0.05;
        }

        if (tempCelsius > 40.0 || tempCelsius < 10.0) {
            impactFactor += 0.08;
        }

        return currentPrice * impactFactor;
    }
};

// 2. Distance Sorting Logic for Thanda Ghar
class StorageUnit {
public:
    string name;
    double latitude;
    double longitude;

    StorageUnit(string n, double lat, double lon) : name(n), latitude(lat), longitude(lon) {}
};

// Haversine formula implementation
double haversineCalculateDistance(double lat1, double lon1, double lat2, double lon2) {
    const double R = 6371.0; // Radius of the earth in km
    const double TO_RAD = M_PI / 180.0;
    
    double latDistance = (lat2 - lat1) * TO_RAD;
    double lonDistance = (lon2 - lon1) * TO_RAD;
    
    double a = sin(latDistance / 2) * sin(latDistance / 2) + 
               cos(lat1 * TO_RAD) * cos(lat2 * TO_RAD) * 
               sin(lonDistance / 2) * sin(lonDistance / 2);
               
    double c = 2 * atan2(sqrt(a), sqrt(1 - a));
    return R * c;
}

// Sorts list of Thanda Ghar units
void sortStorageLocators(double farmerLat, double farmerLon, vector<StorageUnit>& units) {
    sort(units.begin(), units.end(), [farmerLat, farmerLon](const StorageUnit& a, const StorageUnit& b) {
        double distA = haversineCalculateDistance(farmerLat, farmerLon, a.latitude, a.longitude);
        double distB = haversineCalculateDistance(farmerLat, farmerLon, b.latitude, b.longitude);
        return distA < distB;
    });
}

int main() {
    // Example Usage
    CropPricing wheat("Wheat", 2200.0);
    double predicted = wheat.calculatePredictedPrice(160.0, 35.0, false);
    cout << "Predicted Wheat Price: " << predicted << endl;

    vector<StorageUnit> storages = {
        StorageUnit("Kisan Storage", 28.7041, 77.1025),
        StorageUnit("AgriCold", 28.5355, 77.3910)
    };
    
    // Sort and output nearby units
    sortStorageLocators(28.6139, 77.2090, storages);
    for (const auto& s : storages) {
        cout << "Storage: " << s.name << endl;
    }

    return 0;
}
